proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,111 @@
+* Support immutable failure handling policy configuration
+* Some Failure policy configurations, e.g. CircuitBreaker, RetryPolicy, can be used stand alone. For example, it has been very useful for circuit breakers to be standalone constructs which can be plugged into and intentionally shared across multiple executions. Likewise for retry policies. Additionally, an Execution construct can be offered that allows retry policies to be applied to some logic in a standalone, manually controlled way.
+
+Advanced requirements:
+
+* Event Hooks: Since this approach to fault tolerance involves handing execution over to some foreign code, it's very useful to be able to learn when executions are taking place and under what circunstances (onRetry, onFailedAttempt, onFailure, etc).
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/ezFC1TLGozU)
+
+## Motivation
+
+Currently there are at least two libraries to provide fault tolerance. It is best to uniform the technologies and define a standard so that micro service applications can adopt and the implementation of fault tolerance can be provided by the containers if possible.
+
+## Proposed solution
+
+Separate the responsibility of executing logic (Runnables/Callables/etc) from guiding when execution should take place (through retry policies, bulkheads, circuit breakers). In this way, failure handling strategies become configuration that can influence executions, and the execution API itself is just responsible for receiving some configuration and performing executions.
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* A Java API to privide RetryPolicy
+* A Java API to provide FallBack
+* A Java API to provide BulkHead
+* A Java API to provide CircuitBreaker
+
+## Detailed design
+
+The Design is built on CDI programming model. An instance of `Execution`, `RetryPolicy`, `CircuitBreaker` and `BulkHeader` can be injected to the clients and then they can be configured as follows.
+
+### RetryPolicy: A policy define the retry criteria
+
+
+```
+RetryPolicy rp = retryPolicy.retryOn(TimeOutException.class)
+  .withDelay(2, TimeUnit.SECONDS)
+  .withMaxRetries(2);
+```
+
+When `TimeOutException` was received, delay 2 seconds and then retry 2 more times.
+
+```
+Connection connection = execution.with(rp).withFallBack(null).get(this::connect)
```

Comment:
@jhalterman So withFallback is a T rather than () => T (which I am proposing should rather be a () => Optional[T]) I.e. In case there's a problem, here's how you obtain a replacement. If the strategy is to rely on `null` to signal absence (which I'd consider an antipattern) or forcing the fallback to throw exceptions in case of inability to fall back (also an antipattern imo) I'd rather have exceptions signal truly exceptional cases and having a function to produce the fallback value means that they do not need to be created if never needed.
Does that make more sense? :)